src/production/production_16.py

```diff
@@ -0,0 +1,53 @@
+import itertools as it
+import matplotlib.pyplot as plt
+import util
+import basic_graph
+from typing import Dict
+from model import NodeAttrs, EdgeAttrs, NodeHandle, Edge, Node
+from graph import Graph
+from .model import Production
+
+class P16(Production):
+    def __init__(self, idx=None) -> None:
+        self.lhs: Graph = self.__create_lhs()
+        self.rev_mapping: Dict[NodeHandle, NodeHandle] | None = None
+        self.idx: NodeHandle = idx
+
+    def get_lhs(self) -> Graph:
+        return self.lhs
+
+    def requires_monomorphism(self):
+        return True
+
+    def reset(self):
+        self.rev_mapping = None
+
+    def __create_lhs(self) -> Graph:
+#       graph, _ = util.basic_star5(for_lhs=True, select_central=False)
+        graph, _ = basic_graph.basic_star5(for_lhs=True, select_central=False)
+        return graph
+
+    def apply_with_mapping(self, graph: Graph, mapping: Dict[NodeHandle, NodeHandle]):
+        # Break the edges (1, 2), (2, 3), (3, 4), (4, 1), creating 4 new nodes
+        # Add 5 new node to the center of the split
+        self.rev_mapping = util.reverse_dict_mapping(mapping)
+
+        # Aliasing for convenience
+        rev_mapping = self.rev_mapping
+
+        # get square's vertices
+        vertices = [
+            graph.node_for_handle(rev_mapping[i]) for i in range(0, 5)
+        ]
+        # get inner node, the hyperedge
+        node_p = graph.node_for_handle(rev_mapping[5])
+        x = node_p.attrs.x
+        y = node_p.attrs.y
+
+        graph.remove_p_hyperedge(node_p.handle)
+
+        attr = EdgeAttrs('p', flag=True)
+        graph.add_p_hyperedge(vertices, attr, node_p.handle, (x, y)) 
```

Comment:
I didn't reason much what's going on here, but if you only removing the p edge, changing flag value & reinserting it into the same place: you can just edit `node_p.attrs.flag = True` without adding/removing the edge and it should work as you get reference to original attributes.

If it is not the case, just ignore this comment.

Reply:
This is exactly the case, and I'm almost sure I tried this approach and failed for some unknown reason so it's why this spaghetti code is present. Nevertheless, now it's perfectly working and refactoring code in this way